### ISSUE TITLE
Update docker registry package

### ISSUE THREAD

**Issue by kargakis** [openshift/origin#4339]
https://github.com/docker/distribution/releases

When that time comes, [code.google.com/p/go-uuid/uuid](https://github.com/openshift/origin/tree/master/Godeps/_workspace/src/code.google.com/p/go-uuid/uuid) which is a duplicate of [https://github.com/pborman/uuid](https://github.com/openshift/origin/tree/master/Godeps/_workspace/src/github.com/pborman/uuid) can be removed from our deps.

**Comment by kargakis** [comment]
> When that time comes, code.google.com/p/go-uuid/uuid which is a duplicate of https://github.com/pborman/uuid can be removed from our deps.

Of course some imports in our codebase need to be changed

**Comment by ncdc** [comment]
We can't bump docker/distribution until we refactor image pruning (#3333) due to some hacks we made to the distribution Godep that are very difficult to carry forward with the latest upstream code.

**Comment by smarterclayton** [comment]
Do we have adequate attention on making sure this lands as fast as is possible?  This is now a serious blocking issue for a huge number of issues.

**Comment by ncdc** [comment]
We still need an upstream PR merged before we can proceed, plus code review on the origin changes.

**Comment by smarterclayton** [comment]
Timeframe on upstream pr?

**Comment by ncdc** [comment]
I just asked for an update; not sure at this point.

**Comment by pweil-** [comment]
linking the upstream: https://github.com/docker/distribution/pull/1050

**Comment by pweil-** [comment]
https://github.com/docker/distribution/pull/1050 is still under review.  It may have a chance at getting merged this week but I don't think we can guarantee it.  Given that this is a P0 should we proceed with trying to carry the pruning hacks and resolving this or continue to wait on the upstream PR?

**Comment by ncdc** [comment]
:-1: to carrying hacks - we've learned our lesson the hard way :cry:

**Comment by smarterclayton** [comment]
We are falling really far behind on the registry.  I think we can't afford
to keep falling behind.  Can we asses the risk of the rebase in terms of
possibility of breakage of end user workflows?

On Dec 8, 2015, at 10:33 AM, Andy Goldstein <notifications@github.com>
wrote:

[image: :-1:] to carrying hacks - we've learned our lesson the hard way [image:
:cry:]

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/4339#issuecomment-162917099>.

**Comment by ncdc** [comment]
The only thing we have to deal with is pruning and possibly the custom
/healthz handler I hacked in. I believe the latest registry code responds
200 at / now, which is a sufficient replacement for /healthz. So that just
leaves pruning, which depends on upstream 1050 which isn't in yet.

**Comment by miminar** [comment]
> I believe the latest registry code responds 200 at / now

Correct.

> So that just leaves pruning, which depends on upstream 1050 which isn't in yet.

I'd rather rework pruning after the rebase itself (with or without 1050 carried on top). Doing it together would result in unreviewable PR like the last time. We could rebase on top of `registry:v2.2.0` or [`registry:2.2.1`](https://github.com/docker/distribution/issues?q=milestone%3ARegistry%2F2.2.1+is%3Aclosed)  which is due by December 9, 2015.

 Other note-able changes:
- repository name components can now be one character long
- error messages were reworked
- new features ([v2.1](https://github.com/docker/distribution/issues?utf8=%E2%9C%93&q=milestone%3ARegistry%2F2.1+label%3Afeature+), [v.2.2](https://github.com/docker/distribution/issues?utf8=%E2%9C%93&q=milestone%3ARegistry%2F2.2+label%3Afeature)) like support for repo listing and pagination, new backend drivers, pull-through-cache, communication via unix-socket

Those shouldn't break existing workflows.

**Comment by miminar** [comment]
Upstream wants to refactor manifest interface before accepting anything else, which means another not-so-short delay for 1050.

**Comment by smarterclayton** [comment]
Which means... We have to carry 1050 if we want to upgrade docker?  Right
now it seems we have all the cost without even getting a new registry
version.  Do we need to release the registry image on a different window
than 3.1.1 in order to get enough time to get this work done?

On Dec 10, 2015, at 3:09 AM, Michal Minar <notifications@github.com> wrote:

Upstream wants to refactor manifest interface before accepting anything
else, which means another not-so-short delay for 1050.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/issues/4339#issuecomment-163529319>.

**Comment by miminar** [comment]
Yes, we have to carry 1050 if we want to rebase the registry. I don't think 1050 makes it before 3.1.1. Last time I asked upstream, they told me they want to merge it before refactoring. Now they change their mind. So it's really hard to guess. I'll ask them on ETA of the refactoring.

**Comment by smarterclayton** [comment]
Can you make 1050 carry + rebase happen for 3.1.1?  Risk profile sounded
less than I was worried about.

**Comment by miminar** [comment]
@smarterclayton Sure. Working on it.

**Comment by miminar** [comment]
@smarterclayton: Rebased docker/distribution (PR #5404) resurrected.
